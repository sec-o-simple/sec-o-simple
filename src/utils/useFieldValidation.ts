import useValidationStore from '@/utils/useValidationStore'

export function useFieldValidation(path?: string) {
  const validationStore = useValidationStore()

  const messages = !path ? [] : validationStore.getMessagesForPath(path)
  const errorMessages = messages.filter((m) => m.severity === 'error')
  const warningMessages = messages.filter((m) => m.severity === 'warning')
  const infoMessages = messages.filter((m) => m.severity === 'info')

  return {
    messages,
    hasErrors: errorMessages.length > 0,
    hasWarnings: warningMessages.length > 0,
    hasInfos: infoMessages.length > 0,
    errorMessages,
    warningMessages,
    infoMessages,
    isTouched: !path ? true : validationStore.isFieldTouched(path),
    markFieldAsTouched: validationStore.markFieldAsTouched,
  }
}
