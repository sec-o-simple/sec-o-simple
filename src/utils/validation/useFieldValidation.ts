import useValidationStore from '@/utils/validation/useValidationStore'

export function useFieldValidation(path?: string) {
  const validationStore = useValidationStore()

  const messages = !path ? [] : validationStore.getMessagesForPath(path)
  const errorMessages = messages.filter((m) => m.severity === 'error')
  const warningMessages = messages.filter((m) => m.severity === 'warning')
  const infoMessages = messages.filter((m) => m.severity === 'info')

  return {
    messages,
    hasErrors: errorMessages.length > 0,
    hasWarnings: warningMessages.length > 0,
    hasInfos: infoMessages.length > 0,
    errorMessages,
    warningMessages,
    infoMessages,
    // For now we mark all fields as touched until we have a better way to handle this
    isTouched: true,
    markFieldAsTouched: validationStore.markFieldAsTouched,
  }
}
