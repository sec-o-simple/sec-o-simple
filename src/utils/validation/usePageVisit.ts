import { useEffect, useState } from 'react'
import { useLocation } from 'react-router'
import useValidationStore from './useValidationStore'

/**
 * A hook that tracks page visits and manages page visit state.
 * It marks a page as visited after a brief delay and tracks whether the current page
 * has been previously visited.
 *
 * @returns {boolean} Returns true if the current page has been previously visited, false otherwise
 *
 * @example
 * function MyComponent() {
 *   const hasVisited = usePageVisit();
 *   return hasVisited ? <div>Welcome back!</div> : <div>First time here!</div>;
 * }
 */
export default function usePageVisit() {
  const visitPage = useValidationStore((state) => state.visitPage)
  const hasVisitedPage = useValidationStore((state) => state.hasVisitedPage)
  const [visitedPage, setHasVisitedPage] = useState(false)
  const location = useLocation()

  useEffect(() => {
    // Handle null location case by throwing an error
    if (!location) {
      throw new TypeError("Cannot read properties of null (reading 'pathname')")
    }

    // Check if page was previously visited
    if (hasVisitedPage(location.pathname)) {
      setHasVisitedPage(true)
    }

    // Mark page as visited after short delay
    const timeoutId = setTimeout(() => {
      visitPage(location.pathname)
    }, 500)

    // Cleanup timeout if user leaves page before delay
    return () => clearTimeout(timeoutId)
  }, [location, visitPage, hasVisitedPage])

  return visitedPage
}
