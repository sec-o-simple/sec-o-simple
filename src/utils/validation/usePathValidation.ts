import useValidationStore from './useValidationStore'

type HasErrorFunction = (errorPaths: string[]) => boolean

const validationSections: Record<string, HasErrorFunction> = {
  '/document-information/general': (errorPaths) => {
    return ['/document/title', '/document/tracking/id', '/document/lang'].some(
      (path) => errorPaths.includes(path),
    )
  },
  '/tracking': (errorPaths) => {
    return (
      errorPaths.includes('/document/tracking/status') ||
      errorPaths.some((path) =>
        path.startsWith('/document/tracking/revision_history'),
      )
    )
  },
  '/document-information/notes': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/document/notes'))
  },
  '/document-information/publisher': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/document/publisher'))
  },
  '/document-information/acknowledgments': (errorPaths) => {
    return errorPaths.some((path) =>
      path.startsWith('/document/acknowledgments'),
    )
  },
  '/document-information/references': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/document/references'))
  },
  '/products': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/products'))
  },
  '/vulnerabilities': (errorPaths) => {
    return errorPaths.some((path) => path.startsWith('/vulnerabilities'))
  },
}

export function usePathValidation(path: string) {
  const messages = useValidationStore((state) => state.messages)
  const errorPaths = messages
    .filter((m) => m.severity === 'error')
    .map((e) => e.path)
  // const hasVisitedPage = useValidationStore((state) => state.hasVisitedPage)
  if (validationSections[path]) {
    const hasErrors = validationSections[path](errorPaths)
    return {
      hasErrors,
      // For now we mark all fields as touched until we have a better way to handle this
      // hasVisited: hasVisitedPage(path),
      hasVisited: true,
    }
  }

  return {
    hasErrors: false,
    // For now we mark all fields as touched until we have a better way to handle this
    // hasVisited: hasVisitedPage(path),
    hasVisited: true,
  }
}
