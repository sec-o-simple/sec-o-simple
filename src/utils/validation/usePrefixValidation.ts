import useValidationStore from "./useValidationStore"

export function usePrefixValidation(prefix: string) {
    const messages = useValidationStore((state) => state.messages)
    const errorPaths = messages
        .filter((m) => m.severity === 'error')
        .map((e) => e.path)

    const hasErrors = errorPaths.some((path) => path.startsWith(prefix + '/'))

    return {
        hasErrors,
    }
}