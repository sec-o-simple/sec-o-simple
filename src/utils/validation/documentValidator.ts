import { TDocumentStore } from '@/utils/useDocumentStore'
import { createCSAFDocument } from '../csafExport/csafExport'

import * as basic from '@secvisogram/csaf-validator-lib/basic.js'
import validate from '@secvisogram/csaf-validator-lib/validate.js'
import { TConfig } from '../useConfigStore'
import { ValidationMessage } from './useValidationStore'

const tests: unknown[] = [...Object.values(basic)]

export interface ValidationResult {
  isValid: boolean
  messages: ValidationMessage[]
}

export async function validateDocument(
  documentStore: TDocumentStore,
  getRelationshipFullProductName: (
    sourceVersionId: string,
    targetVersionId: string,
    category: string,
  ) => string,
  config?: TConfig,
): Promise<ValidationResult> {
  try {
    const csafDocument = createCSAFDocument(
      documentStore,
      getRelationshipFullProductName,
      config,
    )
    const result = await validate(tests, csafDocument)

    // Collect all messages with their severity
    let messages: ValidationMessage[] = []

    result.tests.forEach((test) => {
      test.errors.forEach((error) => {
        messages.push({
          path: error.instancePath,
          message: error.message,
          severity: 'error',
        })
      })

      test.warnings.forEach((warning) => {
        messages.push({
          path: warning.instancePath,
          message: warning.message,
          severity: 'warning',
        })
      })

      test.infos.forEach((info) => {
        messages.push({
          path: info.instancePath,
          message: info.message,
          severity: 'info',
        })
      })
    })

    return {
      // We want to use result.isValid when we implemented all fields
      // but for now we use the messages length to determine if the document is valid
      isValid: messages.filter((m) => m.severity === 'error').length === 0,
      messages,
    }
  } catch (error) {
    return {
      isValid: false,
      messages: [],
    }
  }
}
