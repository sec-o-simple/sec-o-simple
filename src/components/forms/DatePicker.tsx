import { useFieldValidation } from '@/utils/validation/useFieldValidation'
import {
  DatePicker as HeroUIDatePicker,
  DatePickerProps as HeroUIDatePickerProps,
} from '@heroui/date-picker'
import { getLocalTimeZone, parseAbsolute } from '@internationalized/date'

export type DatePickerProps = Omit<
  HeroUIDatePickerProps,
  'value' | 'onChange'
> & {
  csafPath?: string
  isTouched?: boolean
  value?: string
  onChange?: (date: string) => void
}

export default function DatePicker({
  labelPlacement,
  variant,
  csafPath,
  isTouched,
  onBlur,
  onChange,
  value,
  ...pickerProps
}: DatePickerProps) {
  const validation = useFieldValidation(csafPath)

  return (
    <HeroUIDatePicker
      variant={variant ?? 'bordered'}
      labelPlacement={labelPlacement ?? 'outside'}
      classNames={{
        inputWrapper: 'border-1 shadow-none',
      }}
      errorMessage={validation.errorMessages.map((m) => m.message).join(', ')}
      isInvalid={
        validation.hasErrors &&
        (isTouched || validation.isTouched || !!value?.length)
      }
      value={value ? parseAbsolute(value, getLocalTimeZone()) : undefined}
      onBlur={(e) => {
        if (csafPath) {
          validation.markFieldAsTouched(csafPath)
        }
        onBlur?.(e)
      }}
      onChange={(dateValue) =>
        onChange?.(dateValue?.toDate(getLocalTimeZone()).toISOString() ?? '')
      }
      {...pickerProps}
    />
  )
}
