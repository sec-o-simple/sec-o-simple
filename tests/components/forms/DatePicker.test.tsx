import { render, screen, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import '@testing-library/jest-dom'
import DatePicker from '../../../src/components/forms/DatePicker'
import { useFieldValidation } from '../../../src/utils/validation/useFieldValidation'
import { getLocalTimeZone, parseAbsolute } from '@internationalized/date'

// Mock the validation hook
vi.mock('../../../src/utils/validation/useFieldValidation', () => ({
  useFieldValidation: vi.fn(() => ({
    messages: [],
    errorMessages: [],
    warningMessages: [],
    infoMessages: [],
    hasErrors: false,
    hasWarnings: false,
    hasInfos: false,
    isTouched: false,
    markFieldAsTouched: vi.fn(),
  })),
}))

// Mock internationalized date functions
vi.mock('@internationalized/date', () => ({
  getLocalTimeZone: vi.fn(() => 'UTC'),
  parseAbsolute: vi.fn((date) => ({
    toDate: vi.fn(() => new Date(date)),
  })),
}))

// Mock HeroUI DatePicker component
vi.mock('@heroui/date-picker', () => ({
  DatePicker: vi.fn(({ children, errorMessage, isInvalid, classNames, value, onChange, onBlur, labelPlacement, variant, label, ...props }: any) => (
    <input 
      type="date" 
      data-testid="hero-datepicker"
      data-error-message={errorMessage || ''}
      data-is-invalid={isInvalid?.toString() || 'false'}
      data-classnames={typeof classNames === 'object' ? JSON.stringify(classNames) : classNames}
      data-value={value?.toString() || ''}
      data-label-placement={labelPlacement}
      data-variant={variant}
      data-label={label}
      onChange={(e) => {
        if (onChange) {
          const mockDateValue = {
            toDate: vi.fn(() => new Date(e.target.value)),
          }
          onChange(mockDateValue)
        }
      }}
      onBlur={onBlur}
    />
  )),
}))

const mockUseFieldValidation = vi.mocked(useFieldValidation)
const mockParseAbsolute = vi.mocked(parseAbsolute)
const mockGetLocalTimeZone = vi.mocked(getLocalTimeZone)

describe('DatePicker', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    mockUseFieldValidation.mockReturnValue({
      messages: [],
      errorMessages: [],
      warningMessages: [],
      infoMessages: [],
      hasErrors: false,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    })
    mockParseAbsolute.mockImplementation((date) => ({
      toDate: vi.fn(() => new Date(date)),
    }))
    mockGetLocalTimeZone.mockReturnValue('UTC')
  })

  it('should render with default props', () => {
    const { container } = render(<DatePicker />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toBeInTheDocument()
    expect(datePicker).toHaveAttribute('data-variant', 'bordered')
    expect(datePicker).toHaveAttribute('data-label-placement', 'outside')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should accept custom props', () => {
    render(
      <DatePicker
        label="Custom Date"
        variant="underlined"
        labelPlacement="inside"
      />
    )
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-label', 'Custom Date')
    expect(datePicker).toHaveAttribute('data-variant', 'underlined')
    expect(datePicker).toHaveAttribute('data-label-placement', 'inside')
  })

  it('should apply custom classNames for inputWrapper', () => {
    render(<DatePicker />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-classnames', JSON.stringify({
      inputWrapper: 'border-1 shadow-none',
    }))
  })

  it('should handle value prop with ISO string', () => {
    const mockDateValue = {
      toDate: vi.fn(() => new Date('2023-12-25')),
    }
    mockParseAbsolute.mockReturnValue(mockDateValue)
    
    render(<DatePicker value="2023-12-25T00:00:00.000Z" />)
    
    expect(mockParseAbsolute).toHaveBeenCalledWith('2023-12-25T00:00:00.000Z', 'UTC')
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-value', mockDateValue.toString())
  })

  it('should handle undefined value', () => {
    render(<DatePicker />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-value', '')
  })

  it('should handle onChange events', () => {
    const onChange = vi.fn()
    
    render(<DatePicker onChange={onChange} />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    fireEvent.change(datePicker, { target: { value: '2023-12-25' } })
    
    expect(onChange).toHaveBeenCalled()
  })

  it('should mark field as touched on blur with csafPath', async () => {
    const user = userEvent.setup()
    const mockMarkFieldAsTouched = vi.fn()
    const mockValidation = {
      messages: [],
      errorMessages: [],
      warningMessages: [],
      infoMessages: [],
      hasErrors: false,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: mockMarkFieldAsTouched,
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<DatePicker csafPath="test.date" />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    await user.click(datePicker)
    fireEvent.blur(datePicker)
    
    expect(mockMarkFieldAsTouched).toHaveBeenCalledWith('test.date')
  })

  it('should display validation errors when field has errors and is touched', () => {
    const mockValidation = {
      messages: [{ message: 'Invalid date format', severity: 'error' }],
      errorMessages: [{ message: 'Invalid date format' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: true,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<DatePicker csafPath="test.date" />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-error-message', 'Invalid date format')
    expect(datePicker).toHaveAttribute('data-is-invalid', 'true')
  })

  it('should display validation errors when field has errors and isTouched prop is true', () => {
    const mockValidation = {
      messages: [{ message: 'Date is required', severity: 'error' }],
      errorMessages: [{ message: 'Date is required' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<DatePicker csafPath="test.date" isTouched />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-error-message', 'Date is required')
    expect(datePicker).toHaveAttribute('data-is-invalid', 'true')
  })

  it('should display validation errors when field has errors and has value', () => {
    const mockValidation = {
      messages: [{ message: 'Invalid date', severity: 'error' }],
      errorMessages: [{ message: 'Invalid date' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<DatePicker csafPath="test.date" value="2023-12-25T00:00:00.000Z" />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-error-message', 'Invalid date')
    expect(datePicker).toHaveAttribute('data-is-invalid', 'true')
  })

  it('should not display validation errors when field has errors but no triggering conditions', () => {
    const mockValidation = {
      messages: [{ message: 'Date is required', severity: 'error' }],
      errorMessages: [{ message: 'Date is required' }],
      warningMessages: [],
      infoMessages: [],
      hasErrors: true,
      hasWarnings: false,
      hasInfos: false,
      isTouched: false,
      markFieldAsTouched: vi.fn(),
    }
    
    mockUseFieldValidation.mockReturnValue(mockValidation)
    
    render(<DatePicker csafPath="test.date" />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    expect(datePicker).toHaveAttribute('data-error-message', 'Date is required')
    expect(datePicker).toHaveAttribute('data-is-invalid', 'false')
  })

  it('should work without csafPath (no validation)', () => {
    const onChange = vi.fn()
    
    render(<DatePicker onChange={onChange} />)
    
    const datePicker = screen.getByTestId('hero-datepicker')
    fireEvent.change(datePicker, { target: { value: '2023-12-25' } })
    
    expect(onChange).toHaveBeenCalled()
    expect(datePicker).toHaveAttribute('data-error-message', '')
    expect(datePicker).toHaveAttribute('data-is-invalid', 'false')
  })

  describe('Snapshots', () => {
    it('should render default state', () => {
      const { container } = render(<DatePicker />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should render with error state', () => {
      const mockValidation = {
        messages: [{ message: 'Required field', severity: 'error' }],
        errorMessages: [{ message: 'Required field' }],
        warningMessages: [],
        infoMessages: [],
        hasErrors: true,
        hasWarnings: false,
        hasInfos: false,
        isTouched: true,
        markFieldAsTouched: vi.fn(),
      }
      
      mockUseFieldValidation.mockReturnValue(mockValidation)
      
      const { container } = render(<DatePicker csafPath="test.date" />)
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should render with value', () => {
      const { container } = render(<DatePicker value="2023-12-25T00:00:00.000Z" />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
