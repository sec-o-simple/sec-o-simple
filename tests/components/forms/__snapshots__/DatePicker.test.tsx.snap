// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatePicker > Snapshots > should render default state 1`] = `
<input
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-error-message=""
  data-is-invalid="false"
  data-label-placement="outside"
  data-testid="hero-datepicker"
  data-value=""
  data-variant="bordered"
  type="date"
/>
`;

exports[`DatePicker > Snapshots > should render with error state 1`] = `
<input
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-error-message="Required field"
  data-is-invalid="true"
  data-label-placement="outside"
  data-testid="hero-datepicker"
  data-value=""
  data-variant="bordered"
  type="date"
/>
`;

exports[`DatePicker > Snapshots > should render with value 1`] = `
<input
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-error-message=""
  data-is-invalid="false"
  data-label-placement="outside"
  data-testid="hero-datepicker"
  data-value="[object Object]"
  data-variant="bordered"
  type="date"
/>
`;

exports[`DatePicker > should render with default props 1`] = `
<input
  data-classnames="{"inputWrapper":"border-1 shadow-none"}"
  data-error-message=""
  data-is-invalid="false"
  data-label-placement="outside"
  data-testid="hero-datepicker"
  data-value=""
  data-variant="bordered"
  type="date"
/>
`;
